File: achilles_sensor/serial/scripts/MultCode.py
#!/usr/bin/env python         
# -*- coding: utf-8 -*-

class MultCode:

    #各个字段的长度
    __fromeSize = 2
    __toSize = 2
    __numSize = 8
    __typeSize = 4
    __dataSize = 8
    __sumSize = __fromeSize + __toSize + __numSize + __typeSize + __dataSize

    def __init__(self):
        '初始化'
        #发送方
        self.__from = 0
        
        #接受方
        self.__to = 0

        #设备编号
        self.__num = 0

        #设备中的某个数据的类型编号
        self.__type = 0

        #数据
        self.__data = 0

        
        #保存传入数据
        self.__rxdata = " "
        #保存传出数据
        self.__txdata = " "

        #缓存区
        self.__tempBuffer = [" " for _ in range(self.__sumSize)]

        print ("协议转换器初始化完成")

    def __judgement(self, data, bit):
        '数据 位数'
        s = str(data)
        compare = 1
        for i in range(1,bit):
            compare = compare * 10
        
        for i in range(1,bit):
            if compare > data:
                s = "0" + s
            compare = compare / 10
        return s
    

    def __cut(self, begin, end):
        s = ""
        for i in range(begin, end):
            s = s + self.__tempBuffer[i]
        return s

    def setFrom(self, data):
        '设置 from 字段'
        self.__from = data
    
    def setTo(self, data):
        '设置 to 字段'
        self.__to = data
    
    def setNum(self, data):
        '设置 num 字段'
        self.__num = data

    def setType(self, data):
        '设置 type 字段'
        self.__type = data

    def setData(self, data):
        '设置 data 字段'
        self.__data = data

    def getFrom(self):
        '获取 from 字段'
        return self.__from

    def getTo(self):
        '获取 to 字段'
        return self.__to

    def getNum(self):
        '获取 num 字段'
        return self.__num

    def getType(self):
        '获取 type 字段'
        return self.__type

    def getData(self):
        '获取 data 字段'
        return self.__data
    
    def getSumSize(self):
        '获取标准协议总长度'
        return self.__sumSize
    


    def encode(self, from_, to_, num_, type_, data_):
        '不同字段的长度： form为2 to为2 num为8 type为4 data为8'
        _from = self.__judgement(from_, self.__fromeSize)
        _to = self.__judgement(to_, self.__toSize)
        _num = self.__judgement(num_, self.__numSize)
        _type = self.__judgement(type_, self.__typeSize)
        _data = self.__judgement(data_, self.__dataSize)
        
        self.__txdata = _from + _to + _num + _type + _data

        #print (self.__txdata)
        return self.__txdata

    def decode(self, data_):
        #print ("decoding...")
        self.__rxdata = data_
        self.__tempBuffer = list(data_)
        #print self.__tempBuffer
        position = 0
        
        self.setFrom(int(self.__cut(position, position + self.__fromeSize)))
        position += self.__fromeSize

        self.setTo(int(self.__cut(position, position + self.__toSize)))
        position += self.__toSize

        self.setNum(int(self.__cut(position, position + self.__numSize)))
        position += self.__numSize

        self.setType(int(self.__cut(position, position + self.__typeSize)))
        position += self.__typeSize

        self.setData(int(self.__cut(position, position + self.__dataSize)))
        position += self.__dataSize
        

        #print ("decode finished")

File: Arduino/libraries/MultCode/MultCode.cpp
#include "MultCode.h"


MultCode::MultCode()
{
    from = 0;
    to = 0;
    num = 0;
    type = 0;
    data = 0;

    rxdata = new char[sumSize];
    txdata = new char[sumSize];
    tempBuffer = new char[sumSize];
    //rxdata = ( char* )malloc( sizeof(char) * sumSize );
    //txdata = ( char* )malloc( sizeof(char) * sumSize );
    //tempBuffer = ( char* )malloc( sizeof(char) * sumSize );
    //cout << "协议转换器初始化完成" << endl;
}

/*
MultCode::~MultCode()
{
    //dtor
}
*/

//输入字段，返回合并后的字段数组
char* MultCode::encode(int from, int to, int num, int type, int data)
{
    char * _from = judgement(from, fromeSize);
    char * _to = judgement(to, toSize);
    char * _num = judgement(num, numSize);
    char * _type = judgement(type, typeSize);
    char * _data = judgement(data, dataSize);

    int position = 0;
    for(int i = 0; i < fromeSize; i ++)
    {
        txdata[position] = _from[i];
        position ++;
    }

    for(int i = 0; i < toSize; i ++)
    {
        txdata[position] = _to[i];
        position ++;
    }

    for(int i = 0; i < numSize; i ++)
    {
        txdata[position] = _num[i];
        position ++;
    }

    for(int i = 0; i < typeSize; i ++)
    {
        txdata[position] = _type[i];
        position ++;
    }

    for(int i = 0; i < dataSize; i ++)
    {
        txdata[position] = _data[i];
        position ++;
    }
    delete _from;
    delete _to;
    delete _num;
    delete _type;
    delete _data;

    return txdata;
}


//输入数组，将不同字段放入不同的字段中。通过get～()方法即可获得数据
void MultCode::decode(char * data)
{
    for(int i = 0 ; i < sumSize; i ++)
    {
        rxdata[i] = data[i];
    }
    for(int i = 0 ; i < sumSize; i ++)
    {
        tempBuffer[i] = data[i];
    }

    int position = 0;


    char * _from = cut(position, position + fromeSize);
    setFrom
    (
        char_ToInt
        (
            _from,
            fromeSize
        )
    );
    position += fromeSize;
    delete _from;


    char * _to = cut(position, position + toSize);
    setTo
    (
        char_ToInt
        (
            _to,
            toSize
        )
    );
    position += toSize;
    delete _to;

    char * _num = cut(position, position + numSize);
    setNum
    (
        char_ToInt
        (
            _num,
            numSize
        )
    );
    position += numSize;
    delete _num;

    char * _type = cut(position, position + typeSize);
    setType
    (
        char_ToInt
        (
            _type,
            typeSize
        )
    );
    position += typeSize;
    delete _type;


    char * _data = cut(position, position + dataSize);
    setData
    (
        char_ToInt
        (
            _data,
            dataSize
        )
    );
    position += dataSize;
    delete _data;
}


void MultCode::setFrom( int data ){this->from = data;}
void MultCode::setTo( int data ){this->to = data;}
void MultCode::setNum( int data ){this->num = data;}
void MultCode::setType( int data ){this->type = data;}
void MultCode::setData( int data ){this->data = data;}

int MultCode::getFrom() const{ return from;}
int MultCode::getTo() const{ return to;}
int MultCode::getNum() const{ return num;}
int MultCode::getType() const{ return type;}
int MultCode::getData() const{ return data;}

int MultCode::getFromeSize() const{return fromeSize;}
int MultCode::getToSize() const{return toSize;}
int MultCode::getNumSize() const{return numSize;}
int MultCode::getTypeSize() const{return typeSize;}
int MultCode::getDataSize() const{return dataSize;}
int MultCode::getSumSize() const{return sumSize;}
char* MultCode::getTxData() const{return txdata;};
char* MultCode::getRxData() const{return rxdata;};

int MultCode::char_ToInt(char* str, int length) const
{
    int num = 0;
    char char_temp;
    int int_temp;
    for(int i = 0 ; i < length; i ++)
    {
        char_temp = str[i];
        int_temp = (int)char_temp -48;
        num += int_temp;
        num *= 10;
    }
    num = num / 10;
    return num;
}

//数据 位数
char* MultCode::judgement(int data, int bit)
{
    //char *s = (char *)malloc(bit * sizeof(char));
    char *s = new char[bit];

    for( int i = 0 ; i < bit ; i++)
    {
        s[i] = '0';
    }

    int remind = 0;

    for(int i = 0; data > 0; i ++)
    {
        remind = data % 10;
        data = data / 10;

        s[bit - i - 1] = (char)(remind + 48);
    }
    return s;
}

//起始位置 结束位置
char* MultCode::cut(int begin, int end)
{
    //char *s = (char *)malloc((end - begin + 1) * sizeof(char));
    char *s = new char[end - begin + 1];
    int j = 0;
    for(int i = begin; i < end; i ++)
    {
        s[j] = this->tempBuffer[i];
        j ++;
    }
    return s;
}
